cpu/gd32v/periph/adc.c

```diff
@@ -0,0 +1,207 @@
+/*
+ * Copyright (C) 2016 Engineering-Spirit
+ *               2023 Gunar Schorcht
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @ingroup     cpu_gd32v
+ * @ingroup     drivers_periph_adc
+ * @{
+ *
+ * @file
+ * @brief       Low-level ADC driver implementation
+ *
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ * @author      Nick van IJzendoorn <nijzendoorn@engineering-spirit.nl>
+ * @author      Gunar Schorcht <gunar@schorcht.net>
+ *
+ * @}
+ */
+
+#include "cpu.h"
+#include "macros/units.h"
+#include "mutex.h"
+#include "periph/adc.h"
+#include "periph_conf.h"
+
+/**
+ * @brief   Maximum allowed ADC clock speed
+ */
+#define MAX_ADC_SPEED           (MHZ(14))
+
+/**
+ * @brief   Allocate locks for all three available ADC devices
+ */
+static mutex_t locks[] = {
+#if ADC_DEVS > 1
+    MUTEX_INIT,
+#endif
+#if ADC_DEVS > 2
+    MUTEX_INIT,
+#endif
+    MUTEX_INIT
+};
+
+static inline ADC0_Type *dev(adc_t line)
+{
+    return (ADC0_Type *)(ADC0_BASE + (adc_config[line].dev << 10));
```

Comment:
I think that's a bit optimistic if more family members are added, there is no grantee the ADCs will all be continuous in memory.

```suggestion
    switch (adc_config[line].dev) {
    case 0:
        return ADC0_BASE;
#if ADC_DEVS > 1
    case 1:
        return ADC1_BASE;
#endif
#if ADC_DEVS > 2
    case 2:
        return ADC2_BASE;
#endif
    default:
        assert(0);
        return NULL;
    }
```

Reply:
That's the way how STM32F1 dealt with it. The question is whether there will ever be more families for which this code could be used. It is more likely that other families will need their own ADC low level drivers like for STM32. At the moment I wouldn't allow the case `ADC_DEVS > 2`.

Reply:
> It is more likely that other families will need their own ADC low level drivers like for STM32

The STM32 families don't all *need* their own ADC driver, they pretty much all use the same ADC IP. The ADC driver mess on STM32 is due to a lack of cleanup - and that has been always postponed with the new ADC API on the horizon.